kong/utils_test.go

```diff
@@ -999,3 +1000,241 @@ func TestFillConsumerGroupPluginDefaults(T *testing.T) {
 		})
 	}
 }
+
+func Test_fillConfigRecord(T *testing.T) {
+	assert := assert.New(T)
+
+	client, err := NewTestClient(nil, nil)
+	assert.NoError(err)
+	assert.NotNil(client)
+
+	tests := []struct {
+		name     string
+		schema   gjson.Result
+		config   Configuration
+		expected Configuration
+	}{
+		{
+			name: "fills defaults for all missing fields",
+			schema: gjson.Parse(`{
+				"fields": {
+					"config":
+						{
+							"type": "record",
+							"fields":[
+								{
+									"enabled":{
+										"type":"boolean",
+										"default":true,
+										"required":true
+									}
+								},
+								{
+									"mappings":{
+										"required":false,
+										"type":"array",
+										"elements":{
+											"type":"record",
+											"fields":[
+												{"name":{"type":"string","required":false}},
+												{"nationality":{"type":"string","required":false}}
+											]
+										}
+									}
+								}
+							]
+						}
+					}
+				}`),
+			config: Configuration{
+				"mappings": []interface{}{
+					map[string]interface{}{
+						"nationality": "Ethiopian",
+					},
+				},
+			},
+			expected: Configuration{
+				"enabled": true,
+				"mappings": []any{
+					Configuration{
+						"name":        nil,
+						"nationality": "Ethiopian",
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		T.Run(tc.name, func(t *testing.T) {
+			configSchema, err := getConfigSchema(tc.schema)
+			assert.NoError(err)
+			config := fillConfigRecord(configSchema, tc.config)
+			assert.NotNil(config)
+			if diff := cmp.Diff(config, tc.expected); diff != "" {
+				t.Errorf(diff)
+			}
+		})
+	}
+}
+
+func Test_FillPluginsDefaults(T *testing.T) {
+	RunWhenKong(T, ">=2.6.0 <2.8.1")
```

Comment:
@pmalek 
We wanted to add tests for a default plugin which was impacted by this bug (discussed [here](https://github.com/Kong/go-kong/pull/302#pullrequestreview-1363700741)).

But the schema for this plugin has changed a lot in the newer versions (and we are fetching the schema from Kong in this test), hence I added this version check so that it doesn't fail unnecessarily.
Probably we can handle it differently so that the test is valid for all versions of kong.

Reply:
Got it. Perhaps we could add another test that would work with more recent versions of Kong? We can have those either as separate tests named accordingly or as subtests guarded by the `RunWhenKong()` with appropriate versions range. WDYT?

If we leave it like that what would ensure us that users working with plugins in Gateway 3.2 are getting the correct behavior?

Reply:
We should definitely add tests for the recent versions. 

One way to do this is to add schema json in the test data itself, instead of fetching it from Kong. But looking at a few other tests it seems we are no longer doing that (Any reason for this?). 

Also, this particular test scenario might get removed if the StatsD plugin changes its schema to remove the array of records from its config in newer versions.

Reply:
You're right. In that case: would be a reasonable ask to suggest that we don't rely on the client itself and just test `FillPluginsDefaults` by specifying the schema in the test? This way we're not relying on the Gateway version and we're only concerned about the function that we're testing, not the environment.

Reply:
@pmalek I've added schema to the test.

Also, while adding a few more tests I discovered that default values were not being applied for `nil` array of records (working fine for empty array). Added handling for this as well.